Implement SAML account authentication and borrowing

ray-lee opened the pull request:
**What's this do?**

Implement SAML account authentication and borrowing.

- When logging in using SAML, a WebView is used to present the SSO login page. When login completes, the cookies received by the WebView are dumped out and stored in the account credentials.
- When determining borrowing subtasks to run, the account is now taken into... account.
- A `BorrowSAMLDownload` subtask runs when the account has SAML credentials. It attempts the download, using cookies from the credentials. If the download is the expected content type, it is saved normally. Otherwise, if the download resembles a login page (it is HTML), the book status is set to `DownloadWaitingForExternalAuthentication`, and the subtask completes.
- The view classes from which the download may have been initiated (`CatalogFragmentBookDetail` and `CatalogPagedViewHolder`) listen for the book status change to `DownloadWaitingForExternalAuthentication`, update themselves accordingly, and navigate to `CatalogSAML20Fragment`.
- `CatalogSAML20Fragment` changes the book state to `DownloadExternalAuthenticationInProgress`, and displays a WebView containing the login page. Cookies from the account credentials are set in the WebView. A download listener is set on the WebView. The login is considered to be complete when the download listener is notified that a download needs to start. This indicates that the WebView wants to download a file that it can't view natively, which includes things like epub and pdf. (In theory the user could navigate to something weird just using links in the login page, but I'm hoping that's unlikely.)
- When login completes, the account credentials are updated with the cookies in the WebView (which now should include at least one new cookie from the login). The borrow task is re-attempted, which will now use the updated cookies.

**Why are we doing this? (w/ JIRA link if applicable)**

This fixes [SIMPLY-2833](https://jira.nypl.org/browse/SIMPLY-2833) and subtasks.

**How should this be tested? / Do these changes have associated tests?**

When a library uses SAML auth, the user should be presented with a button to log in using SSO. When tapped, the identity provider's log in page should appear. The user should be able to log in successfully.

When downloading a book from a content provider that requires authentication, the user should be presented with the identity provider's log in page. The user should be able to log in successfully, and the download should complete.

The next time a book is downloaded from the same content provider, the user should not need to log in again.

Unit tests are included for the new borrowing subtask and cookie handling.

**Dependencies for merging? Releasing to production?**

n/a

**Has the application documentation been updated for these changes?**

Not sure, will find out.

**Did someone actually run this code to verify it works?**

@ray-lee ran the Vanilla app.

twaddington commented:
Looks like your branch needs a merge or a rebase onto develop. A few previously merged PRs are showing up in your diff.

ray-lee commented:
Hmm... there are commits by you and Mark on this branch from before I started working on it, but I don't think they've been merged to develop. I can get rid of the merge commits if I rebase, but I'll need help resolving conflicts from the commits before I started working on the branch.

io7m commented on simplified-webview/src/main/java/org/nypl/simplified/webview/WebViewUtilities.kt:
```diff
@@ -0,0 +1,68 @@
+    time: Long
+  ): Long {
+    return (time / 1000L - 11644473600000L)
+  }
```
:smile:

io7m reviewed (approved):
This all looks good to me.

The web view cookie database stuff makes me feel a bit sick, but it's unavoidable as you said.
I'm slightly concerned by how much complexity SAML brings to the codebase. Once this initial deployment is out of the way, we should look into implementing one of the more mobile-friendly SAML workflows.

twaddington commented:
> Hmm... there are commits by you and Mark on this branch from before I started working on it, but I don't think they've been merged to develop.

Yeah, I think what happened was you probably branched this from Mark's SAML branch, which was later rebased on top of some other work before being merged into develop. That caused your branch to diverge and changed the commit hashes so they no longer match what's in your SAML branch.

> I can get rid of the merge commits if I rebase, but I'll need help resolving conflicts from the commits before I started working on the branch.

What I'd probably do is squash your work down into a single commit. Then, create a fresh branch off develop and cherry-pick just your changes into the new branch. You'll still have to resolve any conflicts but that should give you a nice and clean branch.

ray-lee commented:
Ok, I'll do that.

ray-lee commented:
The initial pull request has been squashed and rebased. To that, I've added handling for some additional WebView cookie database versions (I believe this will cover everything we're likely to see in the versions of Android we support), and some unit tests.

twaddington commented:
Thanks Ray, lots of good work here. Will take a look.

twaddington commented on simplified-accounts-json/src/main/java/org/nypl/simplified/accounts/json/AccountAuthenticationCredentialsJSON.kt:
```diff
@@ -49,7 +55,7 @@ object AccountAuthenticationCredentialsJSON {
     authObject.put("authenticationDescription", credentials.authenticationDescription)
 
-    when (credentials) {
+    val ignored = when (credentials) {
```
This is totally fine with me but I think we should consider adopting the `@Exhaustive` annotation in the near future:

https://github.com/cashapp/exhaustive

I have a feeling something like this will eventually be adopted by the language itself. Good discussion of the alternatives here:

https://github.com/cashapp/exhaustive#alternatives-considered

twaddington commented on simplified-books-borrowing/src/main/java/org/nypl/simplified/books/borrowing/internal/BorrowSAMLDownload.kt:
```diff
@@ -0,0 +1,107 @@
+
+class BorrowSAMLDownload private constructor() : BorrowSubtaskType {
+  companion object : BorrowSubtaskFactoryType {
+    val loginPageContentType: MIMEType = MIMEType("text", "html", mapOf())
```
You can drop the explicit type here since it can be inferred by the assignment.

twaddington commented on simplified-books-borrowing/src/main/java/org/nypl/simplified/books/borrowing/internal/BorrowSAMLDownload.kt:
```diff
@@ -0,0 +1,107 @@
+      requestModifier = { properties ->
+        val requestCookies = sortedMapOf<String, String>()
+
+        cookieStore.get(properties.target).forEach { cookie ->
```
This looks nice and clean. Another way to do this is with `associateTo` but in this case, it doesn't simplify things much so, I'd leave it as-is.

```kotlin
cookieStore.get(properties.target)
    .associateTo(requestCookies) { cookie ->
        Pair(cookie.name, cookie.value)
    }
```

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/associate-to.html

twaddington commented on simplified-books-borrowing/src/main/java/org/nypl/simplified/books/borrowing/internal/BorrowSAMLDownload.kt:
```diff
@@ -0,0 +1,107 @@
+    val cookieStore = cookieManager.cookieStore
+
+    credentials.cookies.forEach { accountCookie ->
+      // HttpCookie.parse allows for multiple name/values in one header string, so it returns a
```
Helpful note 👍

twaddington commented on simplified-ui-accounts/src/main/java/org/nypl/simplified/ui/accounts/saml20/AccountSAML20Fragment.kt:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  private fun loadLoginPage() {
+    this.webView.loadUrl(this.constructLoginURI())
```
If you rotate the device while viewing this fragment does the webview restore its state without refetching the page?

twaddington commented on simplified-ui-accounts/src/main/java/org/nypl/simplified/ui/accounts/saml20/AccountSAML20Fragment.kt:
```diff
@@ -0,0 +1,212 @@
+    // Don't care
+  }
+
+  private fun findNavigationController(): AccountNavigationControllerType {
```
I think it's a little cleaner to use a lazy delegate here.

```kotlin
private val navigationController by lazy {
    NavigationControllers.find(
      activity = this.requireActivity(),
      interfaceType = AccountNavigationControllerType::class.java
    )
}
```

twaddington commented on simplified-ui-catalog/src/main/java/org/nypl/simplified/ui/catalog/saml20/CatalogSAML20Fragment.kt:
```diff
@@ -0,0 +1,229 @@
+    this.webView.setDownloadListener(this::onDownloadStart)
+
+    if (this.viewModel.isWebViewClientReady) {
+      this.loadLoginPage()
```
See previous comment about orientation changes.

You might want to skip this initial load if you have a `savedInstanceState`.

twaddington commented on simplified-ui-catalog/src/main/java/org/nypl/simplified/ui/catalog/saml20/CatalogSAML20ViewModel.kt:
```diff
@@ -0,0 +1,216 @@
+
+  val isWebViewClientReady: Boolean
+    get() = (this.webViewClient as CatalogSAML20WebClient).isReady
+}
```
Looks like there's some significant overlap between the catalog and account fragments and view models. Is this some business logic we might consider consolidating in the future?

twaddington commented on simplified-webview/src/main/java/org/nypl/simplified/webview/WebViewCookieDatabase.kt:
```diff
@@ -0,0 +1,115 @@
+        else -> {
+          logger.warn("no reader found for cookie database version {} -- cookies may not be read correctly", version)
+
+          WebViewCookieDatabaseVUnknown(db)
```
Nice 🚀

twaddington reviewed (approved):
Nice and clean as usual! I didn't see anything concerning.

Left a couple comments to consider for follow-on work but we should go ahead and get this merged whenever you're ready 👍

io7m commented:
Merge! Merge! Merge! Merge!

christhepianist commented:
Was hoping you'd say that! :-)
